inc/azure_macro_utils/macro_utils.h

```diff
@@ -78,73 +78,101 @@ MU_IF(X, "true", "false") => "true"
-{                                                                  \
-    if((int)value<0 || (int)value>=MU_COUNT_ARG(__VA_ARGS__))      \
-    {                                                              \
-        /*this is an error case*/                                  \
-        return "NULL";                                             \
-    }                                                              \
-    else                                                           \
-    {                                                              \
-        return MU_C2(enumName, StringStorage)[value];              \
-    }                                                              \
-}                                                                  \
-int MU_C2(enumName, _FromString)(const char* enumAsString, enumName* destination)  \
-{                                                                               \
-    if(                                                                         \
-        (enumAsString==NULL) || (destination==NULL)                             \
-    )                                                                           \
-    {                                                                           \
-        return MU_FAILURE;                                                      \
-    }                                                                           \
-    else                                                                        \
-    {                                                                           \
-        size_t i;                                                               \
-        for(i=0;i<MU_COUNT_ARG(__VA_ARGS__);i++)                                \
-        {                                                                       \
-            if(strcmp(enumAsString, MU_C2(enumName, StringStorage)[i])==0)      \
-            {                                                                   \
-                *destination = (enumName)i;                                     \
-                return 0;                                                       \
-            }                                                                   \
-        }                                                                       \
-        return MU_FAILURE;                                                      \
-    }                                                                           \
-}                                                                               \
+    MU_DEFINE_ENUM_WITHOUT_INVALID(enumName, MU_C2(enumName, _INVALID), __VA_ARGS__) \
+
+#define MU_DEFINE_ENUM_STRINGS_WITHOUT_INVALID(enumName, ...) \
+    const char* MU_C3(MU_, enumName, _ToString)(enumName value) \
+    { \
+        MU_FOR_EACH_1(MU_INTERNAL_DECLARE_ENUM_VAR, __VA_ARGS__) \
+        int enum_last_value = -1; \
```

Comment:
>-1 [](start = 30, length = 2)

seems to be overwrriten at the first expansion of the MU_INTERNAL_ASSIGN_ENUM_VALUE

Reply:
Yep, but it is needed as we add +1 to it in the same line.

---
In reply to: [328351664](https://github.com/Azure/azure-macro-utils-c/pull/30#discussion_r328351664) [](ancestors = 328351664)